pkg/pool-weighted/contracts/WeightedPool.sol

```diff
@@ -23,18 +23,80 @@ import "./BaseWeightedPool.sol";
 contract WeightedPool is BaseWeightedPool {
     using FixedPoint for uint256;
 
+    uint256 private constant _MAX_TOKENS = 20;
+
+    uint256 internal immutable _totalTokens;
```

Comment:
Similar question: This has an `internal` getter, and it's `immutable`, so it can actually be made `private`. Too nitpicky? Same goes for the scaling factors and normalized weights.

Reply:
_totalTokens definitely can be private. The only reason to make the others internal would be if you needed to reference token3 or scalingFactor1 directly from a derived class, which is hard to imagine. You can call the internal methods to access them. So yes, I can make all those private.

Reply:
In BasePool, where the `_scalingFactor`s were originally defined, they were actually private.

Reply:
Anyway, I think the `private` model from `BasePool` makes the most sense for all three of these (`_totalTokens`, `_scalingFactor`s, and `_normalizedWeight`s). And it should be changed not just here in `WeightedPool` but also in `StablePool` and `LiquidityBootstrappingPool`. Let me know if you disagree.

Reply:
There's a case for WeightedPool and LBP (consulted Nico to see if there was a convention). Gist of that was:

Generally you want to keep things as restricted as possible - but, these are base classes that we intend to be easily extensible. So that's the tradeoff. There's no "danger" in making them internal, since they're immutable.

In the case of the WeightedPools, Nico points out that there is a BaseWeightedPool, and anybody doing something that would require special handling of tokens would probably derive from that instead of WeightedPool. So that would argue for making them private.

However... what if you wanted to make a 30- or 40-token pool that had identical behavior to WeightedPool, just more tokens - we talked about this exact case today in the retro. In that case, it would be inefficient to implement _getNormalizedWeights() and _getScalingFactors() if you couldn't see the "base" tokens. You'd have to reserve memory for the full size, call the base class, fill in the first 20 slots from the base class, then continue for the rest.

So that would argue for making everything internal (except perhaps the actual _tokenX variables). If you want the tokens, you get them from the Vault. What do you think of that? private _tokens (and _totalTokens), and scalingFactors/normalizedWeights internal for extensibility?

Probably the same applies to StablePool- and there is no base stable pool above it to derive from

Reply:
Stable Pool does look at _token values in swap callbacks. So maybe those should stay internal

Reply:
Interesting... the `_token`s have actually always been `internal`, even in `BasePool`, so making them `private` instead would constitute an actual change rather than just a return to norm like I was suggesting with the others. I guess part of this discussion hinges on the design decisions that went into the original `BasePool`, and I'm not privy to those (but Nico probably is).

> In that case, it would be inefficient to implement _getNormalizedWeights() and _getScalingFactors() if you couldn't see the "base" values.

I think this argument is quite strong, so perhaps leaving most/all of these `internal` is indeed the best path forward. Maybe we should just revert the most recent commit, or else make `_totalTokens` private and leave everything else alone. Sorry for the unnecessary tangent.

The rest of the PR is good to go from my perspective, so if we can just make one more patch to arrive at a consistent answer to the `private`/`internal` problem across all relevant pool types then I can take one last glance and approve.

Reply:
Ok; I'm leaving _tokenX internal because I don't want to change StablePool, and though I'm skeptical derived classes would need to use them, it's "safe" to have them internal if they're immutable. Latest push has those changes.